lgc/patch/PatchBufferOp.cpp

```diff
@@ -1645,6 +1657,53 @@ Instruction *PatchBufferOp::makeLoop(Value *const loopStart, Value *const loopEn
   return loopCounter;
 }
 
+// =====================================================================================================================
+// Provide well-defined behavior for accessing null descriptor.
+//
+// @param checkInst : The buffer descriptro user.
+void PatchBufferOp::processDescriptorUser(Value *checkInst) {
+  ExtractElementInst *extractInst = nullptr;
+  unsigned descIdx = 0;
+  Instruction *insertPos = nullptr;
+  if (isa<ExtractElementInst>(checkInst)) {
+    extractInst = dyn_cast<ExtractElementInst>(checkInst);
+    insertPos = extractInst;
+  } else {
+    // TODO: store and load instruction is to be handled
+    llvm_unreachable("Not implemented!");
+  }
+
+  if (extractInst) {
+    m_builder->SetInsertPoint(insertPos);
+    Value *bufferDesc = insertPos->getOperand(descIdx);
+
+    // Check whether is a null descriptor
+    Value *descIcmpVec = m_builder->CreateICmpEQ(bufferDesc, Constant::getNullValue(bufferDesc->getType()));
+    Value *isNullDesc = m_builder->CreateExtractElement(descIcmpVec, uint64_t(0));
+    for (unsigned elemIdx = 1; elemIdx != 4; ++elemIdx) {
+      Value *elem = m_builder->CreateExtractElement(descIcmpVec, elemIdx);
+      isNullDesc = m_builder->CreateAnd(isNullDesc, elem);
+    }
+
+    // If accessing a null descriptor, the buffer address offset of a runtime array is adjusted to zero to avoid
+    // negative address offset
+    for (User *const user : extractInst->users()) {
```

Comment:
This seems over-complicated. Why not just conditionally replace the length with 0, instead of finding adds and subs that use it? (Also it is replacing an add with a conditional sub, which is surely wrong.)

Reply:
The extracted value is NUM_RECORDS. The final array length is the result of lshr. If there is an address offset, there is a sub operation before the lshr. Do you mean to find the lshr and conditinally replace it with 0 or orignal value?